rte/mo_util_array.F90

```diff
@@ -15,59 +15,216 @@ module mo_util_array
+    module procedure zero_array_1D, zero_array_3D, zero_array_4D
+  end interface
+
   private
-  public :: any_vals_less_than, any_vals_outside
+  public :: zero_array, any_vals_less_than, any_vals_outside
 contains
+  !-------------------------------------------------------------------------------------------------
+  ! Values less than a floor
+  !-------------------------------------------------------------------------------------------------
+  logical function any_vals_less_than_1D(array, minVal)
+    real(wp), dimension(:), intent(in) :: array
+    real(wp),               intent(in) :: minVal
+
+#ifdef _OPENACC
+    ! Compact version using intrinsics below
+    ! but an explicit loop is the only current solution on GPUs
+    real(wp) :: minValue
+    integer  :: i
+
+    minValue = minVal
+    !$acc parallel loop copyin(array) reduction(min:minValue)
+    do i = 1, size(array,1)
+      minValue = min(array(i), minValue)
+    end do
+    any_vals_less_than_1D = (minValue < minVal)
+#else
+    any_vals_less_than_1D = any(array < minVal)
+#endif
+end function any_vals_less_than_1D
+  !-------------------------------------------------------------------------------------------------
+  logical function any_vals_less_than_2D(array, minVal)
+    real(wp), dimension(:,:), intent(in) :: array
+    real(wp),                 intent(in) :: minVal
+
+#ifdef _OPENACC
+    ! Compact version using intrinsics below
+    ! but an explicit loop is the only current solution on GPUs
+    real(wp) :: minValue
+    integer  :: i, j
+
+    minValue = minVal
+    !$acc parallel loop collapse(2) copyin(array) reduction(min:minValue)
+    do j = 1, size(array,2)
+      do i = 1, size(array,1)
+        minValue = min(array(i,j), minValue)
+      end do
+    end do
+    any_vals_less_than_2D = (minValue < minVal)
+#else
+    any_vals_less_than_2D = any(array < minVal)
+#endif
+  end function any_vals_less_than_2D
 !-------------------------------------------------------------------------------------------------
-  logical function any_vals_less_than(array, minVal)
+  logical function any_vals_less_than_3D(array, minVal)
     real(wp), dimension(:,:,:), intent(in) :: array
     real(wp),                   intent(in) :: minVal
 
-    real(wp) :: minValue
-    integer  :: i, j, k
+#ifdef _OPENACC
+  ! Compact version using intrinsics below
+  ! but an explicit loop is the only current solution on GPUs
+  real(wp) :: minValue
+  integer  :: i, j, k
 
-    ! This could be written far more compactly as
-    !       any_vals_less_than = any(array < minVal)
-    ! but an explicit loop also works on GPUs
+  minValue = minVal
+  !$acc parallel loop collapse(3) copyin(array) reduction(min:minValue)
+  do k = 1, size(array,3)
+    do j = 1, size(array,2)
+      do i = 1, size(array,1)
+        minValue = min(array(i,j,k), minValue)
+      end do
+    end do
+  end do
+  any_vals_less_than_3D = (minValue < minVal)
+#else
+  any_vals_less_than_3D = any(array < minVal)
```

Comment:
So why keep the alternate with the `any` intrinsic around at all? Do you expect a big performance hit from just adopting the loop version on the CPU as well?